openvm/cuda/src/apc_apply_bus.cu

```diff
@@ -0,0 +1,177 @@
+#include <stdint.h>
+#include <assert.h>
+#include <stdio.h>
+#include "primitives/buffer_view.cuh"
+#include "primitives/constants.h"
+#include "primitives/trace_access.h"
+#include "primitives/histogram.cuh"
+#include "expr_eval.cuh"
+
+extern "C" {
+  typedef struct {
+    uint32_t bus_id; // Bus id this interaction targets (matches periphery chip bus id)
+    uint32_t num_args; // Number of argument expressions for this interaction
+    uint32_t args_index_off; // Starting index into the `DevArgSpan` array for this interaction's args. Layout: [mult, arg0, arg1, ...]
+  } DevInteraction;
+}
+
+// Fixed number of bits for bitwise lookup
+static constexpr uint32_t BITWISE_NUM_BITS = 8u;
+
+// Applies bus interactions to periphery histograms for a batch of APC rows
+__global__ void apc_apply_bus_kernel(
+  // APC related
+  const Fp* __restrict__ d_output, // APC trace (column-major)
+  size_t H, // APC trace height (rows)
+  int num_apc_calls, // number of APC calls (rows)
+
+  // Interaction related
+  const uint32_t* __restrict__ d_bytecode, // bytecode for stack-machine expressions
+  size_t bc_len, // bytecode length (u32 words)
+  const DevInteraction* __restrict__ d_interactions, // interactions array
+  size_t n_interactions, // number of interactions
+  const DevArgSpan* __restrict__ d_arg_spans, // argument spans array
+  size_t n_arg_spans, // number of arg spans
+
+  // Variable range checker related
+  uint32_t var_range_bus_id, // variable range checker bus id
+  uint32_t* __restrict__ d_var_hist, // variable range histogram buffer
+  size_t var_num_bins, // variable range histogram bin count
+
+  // Tuple range checker related
+  uint32_t tuple2_bus_id, // 2-tuple range checker bus id
+  uint32_t* __restrict__ d_tuple2_hist, // tuple2 histogram buffer
+  uint32_t tuple2_sz0, // tuple2 size dim0
+  uint32_t tuple2_sz1, // tuple2 size dim1
+
+  // Bitwise related
+  uint32_t bitwise_bus_id, // bitwise lookup bus id
+  uint32_t* __restrict__ d_bitwise_hist // bitwise lookup histogram buffer
+) {
+  const int warp = (threadIdx.x >> 5);
+  const int lane = (threadIdx.x & 31);
+  const int warps_per_block = (blockDim.x >> 5);
+
+  // Each block processes a bus interaction and each thread within a block process an apc call, which evaluates multiple expressions.
+  // TODO: can we parallelize over expression evaluation?
+  for (int base = blockIdx.x * warps_per_block; base < (int)n_interactions; base += gridDim.x * warps_per_block) {
+    int i = base + warp;
```

Comment:
i think `warp` could be added to the `for` loop init clause, using `i` instead of `base` directly in the `for`

Reply:
I adjusted the comments in this file, could you have another look?